teos/src/responder.rs

```diff
@@ -539,144 +394,138 @@ impl chain::Listen for Responder {
+        if let Some(trackers) = self.check_confirmations(txs.keys().cloned().collect(), height) {
+            self.gatekeeper.delete_appointments(trackers, true);
+        }
 
-            // Rebroadcast those transactions that need to
-            let (_, rejected_trackers) = self.rebroadcast(self.get_txs_to_rebroadcast(height));
-            // Delete trackers rejected during rebroadcast
-            let trackers_to_delete_gk = rejected_trackers
-                .iter()
-                .map(|uuid| (*uuid, self.trackers.lock().unwrap()[uuid].user_id))
-                .collect();
-            self.delete_trackers(
-                &rejected_trackers,
-                &self
-                    .gatekeeper
-                    .delete_appointments_from_memory(&trackers_to_delete_gk),
-                DeletionReason::Rejected,
-            );
+        let mut trackers_to_delete = Vec::new();
+        // We might be connecting a new block after a disconnection (reorg).
+        // We will need to update those trackers that have been reorged.
+        if self.coming_from_reorg() {
+            // Handle reorged transactions. This clears `self.reorged_trackers`.
+            if let Some(trackers) = self.handle_reorged_txs(height) {
+                trackers_to_delete.extend(trackers);
+            }
+        }
 
-            // Remove all receipts created in this block
-            self.carrier.lock().unwrap().clear_receipts();
+        // Rebroadcast those transactions that need to
+        if let Some(trackers) = self.rebroadcast_stale_txs(height) {
+            trackers_to_delete.extend(trackers);
+        }
 
-            if self.trackers.lock().unwrap().is_empty() {
-                log::info!("No more pending trackers");
-            }
+        if !trackers_to_delete.is_empty() {
+            self.gatekeeper
+                .delete_appointments(trackers_to_delete, false);
         }
+
+        // Remove all receipts created in this block
+        self.carrier.lock().unwrap().clear_receipts();
     }
 
     /// Handles reorgs in the [Responder].
     fn block_disconnected(&self, header: &BlockHeader, height: u32) {
         log::warn!("Block disconnected: {}", header.block_hash());
+        // Update the carrier and our tx_index.
         self.carrier.lock().unwrap().update_height(height);
         self.tx_index
             .lock()
             .unwrap()
             .remove_disconnected_block(&header.block_hash());
-
-        for tracker in self.trackers.lock().unwrap().values_mut() {
-            // The transaction has been unconfirmed. Flag it as reorged out so we can rebroadcast it.
-            if tracker.status == ConfirmationStatus::ConfirmedIn(height) {
-                tracker.status = ConfirmationStatus::ReorgedOut;
-            }
-        }
+        // And store the reorged transactions to be retried later.
+        // TODO: Not only confirmed trackers need to be marked as reorged, but trackers that hasn't confirmed but their
```

Comment:
> strictly, not <=

Umm, why not `<`, they all might would be reorged.

> flag those as reorged too

But they might not be reorged if the dispute is confirmed long ago (before the chain split). Just to be clear, the issue here is that we gonna end up having some trackers that are marked as `InMempoolSince` so we normally would think their disputes are on-chain confirmed already, but their disputes aren't because of a reorg.

We would need to pull all the `InMempoolSince` trackers and check if their disputes were confirmed in a reorged block.

Reply:
We can add this to the list of things that need to be addressed if you have not yet.